opal/mca/base/mca_base_event.c

```diff
@@ -0,0 +1,648 @@
+        return rc;
+    }
+
+    return mca_base_event_get_by_index_internal ((int)(uintptr_t) tmp, event, invalidok);
+}
+
+int mca_base_event_get_by_fullname (const char *full_name, mca_base_event_t **event)
+{
+    return mca_base_event_get_by_fullname_internal (full_name, event, false);
+}
+
+static int mca_base_event_get_by_name_internal (const char *project, const char *framework, const char *component, const char *name,
+                                                mca_base_event_t **event, bool invalidok)
+{
+    char *full_name;
+    int ret;
+
+    ret = mca_base_var_generate_full_name4 (project, framework, component, name, &full_name);
+    if (OPAL_SUCCESS != ret) {
+        return OPAL_ERROR;
+    }
+
+    ret = mca_base_event_get_by_fullname_internal (full_name, event, invalidok);
+    free (full_name);
+
+    return ret;
+}
+
+int mca_base_event_get_by_name (const char *project, const char *framework, const char *component, const char *name,
+                                mca_base_event_t **event)
+{
+    return mca_base_event_get_by_name_internal (project, framework, component, name, event, false);
+}
+
+int mca_base_registration_get_event (mca_base_event_registration_t *registration, mca_base_event_t **event)
+{
+    if (OPAL_UNLIKELY(NULL == registration)) {
+        return OPAL_ERR_BAD_PARAM;
+    }
+
+    *event = registration->event;
+
+    return OPAL_SUCCESS;
+}
+
+static int mca_base_event_get_by_index_internal (int index, mca_base_event_t **event, bool invalidok)
+{
+    if (index >= event_count) {
+        return OPAL_ERR_VALUE_OUT_OF_BOUNDS;
+    }
+
+    *event = opal_pointer_array_get_item (&registered_events, index);
+
+    /* variables should never be removed per MPI 3.0 § 14.3.7 */
+    assert (*event);
+
+    if (((*event)->event_flags & MCA_BASE_EVENT_FLAG_INVALID) && !invalidok) {
+        *event = NULL;
+        return OPAL_ERR_VALUE_OUT_OF_BOUNDS;
+    }
+
+    return OPAL_SUCCESS;
+}
+
+int mca_base_event_get_by_index (int index, mca_base_event_t **event)
+{
+    return mca_base_event_get_by_index_internal (index, (mca_base_event_t **) event, false);
+}
+
+/***************************************************************************************************/
+
+int mca_base_event_get_count (int *count)
+{
+    *count = event_count;
+    return OPAL_SUCCESS;
+}
+
+int mca_base_event_register (const char *project, const char *framework, const char *component, const char *name,
+                             const char *description, mca_base_var_info_lvl_t verbosity, opal_datatype_t **datatypes,
+                             unsigned long *offsets, size_t num_datatypes, mca_base_var_enum_t *enumerator, int extent, int bind,
+                             int source, uint32_t flags, mca_base_notify_fn_t notify, void *ctx, mca_base_event_t **event_out)
+{
+    int ret, group_index;
+    mca_base_event_t *event;
+
+    /* ensure the caller did not set an invalid flag */
+    assert (!(flags & 0x3f));
```

Comment:
What is 0x3f?  If there's a specific enum that should be used, use that enum -- a hard-coded hex value here seems dangerous (even if you use a `(FLAG_MAX-1)` kind of value).